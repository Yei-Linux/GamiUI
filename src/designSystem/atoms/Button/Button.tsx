import React from "react";
import { IGeneralProps } from "../../../core/domain/interfaces/IGeneralProps";
import { ButtonType } from "../../../core/domain/types";
import { ButtonWrapper } from "./Button.styles";

export interface IButton extends IGeneralProps {
  /**
   * Content Button to show
   */
  children: React.ReactNode;
  /**
   * Button Type
   */
  type: ButtonType;
}

const Button = ({ children, type, ...args }: IButton) => {
  return (
    <ButtonWrapper {...args} typeStyle={type} type="submit">
      {children}
    </ButtonWrapper>
  );
};

Button.defaultProps = {
  type: "primary",
  border: "MEDIUM",
  heigth: "SMALL",
  width: "SMALL",
};

export default Button;
