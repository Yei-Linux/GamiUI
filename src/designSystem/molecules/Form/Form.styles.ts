import styled from "@emotion/styled";
import RichText from "../../atoms/RichText";

export const FormWrapper = styled.div<{}>`
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  background: "white";
  width: 450px;
  padding: 25px;
  border-radius: 25px;
`;

export const FormItemWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 100%;
  margin-bottom: 1rem;
`;

export const FormLabel = styled(RichText)`
  margin-bottom: 1rem;
`;
