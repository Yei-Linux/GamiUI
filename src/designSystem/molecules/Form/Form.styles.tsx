import styled from '@emotion/styled'
import { mixinFlexVariants } from 'styles/mixins/flex'
import RichText from '../../atoms/RichText'

export const Form = styled.form`
  ${mixinFlexVariants({ justifyContent: 'center', alignItems: 'center' })}

  flex-direction: column;
  background: 'white';
  width: auto;
  padding: 25px;
  border-radius: 25px;
`

export const FormItem = styled.div`
  ${mixinFlexVariants({ justifyContent: 'center' })}

  flex-direction: column;
  width: 100%;

  position: relative;

  margin-bottom: 1rem;
  padding-bottom: 20px;
`

export const FormLabel = styled(RichText)`
  margin-bottom: 1rem;
`

export const FormErrorLabel = styled(RichText)`
  margin-left: 10px;
`

export const FormError = styled.div`
  ${mixinFlexVariants({ alignItems: 'center' })}

  color: #e32b2b;
  position: absolute;
  bottom: 0px;
  font-size: 10px;
  font-weight: 600;
`
