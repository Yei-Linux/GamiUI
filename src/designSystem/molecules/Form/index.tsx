export { default } from "./Form";
export { options } from "./constants";
