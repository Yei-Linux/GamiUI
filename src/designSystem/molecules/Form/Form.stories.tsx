import { Meta } from '@storybook/react'
import { stylesControl } from '../../../core/helpers/storybook'

import Form from '.'
import React from 'react'
import Input from '../../atoms/Input'
import Password from '../../atoms/Password'
import Number from '../../atoms/Number'
import TextArea from '../../atoms/TextArea'
import Switch from '../../atoms/Switch'
import Select from '../../atoms/Select'
import Radio from '../../atoms/Radio'
import Button from '../../atoms/Button'

export default {
  title: 'Molecules/Form',
  component: Form,
  args: { ...Form.defaultProps },
  argTypes: {
    ...stylesControl,
  },
} as Meta

export const Basic = () => (
  <Form
    onSubmitForm={(values: any) => {
      console.log(values)
    }}
  >
    <Form.Item label="Nombres" name="names">
      <Input placeholder="Ingresa tus nombres" />
    </Form.Item>
    <Form.Item label="Contraseña" name="password">
      <Password placeholder="Ingresa tu contraseña" />
    </Form.Item>
    <Form.Item label="Celular" name="phone">
      <Number placeholder="Ingresa tu celular" />
    </Form.Item>
    <Form.Item label="Comentario" name="comment">
      <TextArea placeholder="Ingresa tu comentario" />
    </Form.Item>
    <Form.Item label="Activar Notificaciones" name="isActive">
      <Switch defaultChecked={false} />
    </Form.Item>
    <Form.Item label="Elige un pais" name="country">
      <Select placeholder="Type your option">
        <Select.Option value="one">Option1</Select.Option>
        <Select.Option value="two">Option2</Select.Option>
        <Select.Option value="three">Option3</Select.Option>
      </Select>
    </Form.Item>
    <Form.Item label="Elige tu genero" name="gender">
      <Radio>
        <Radio.Item isChecked value="one">
          A
        </Radio.Item>
        <Radio.Item value="two">B</Radio.Item>
        <Radio.Item value="three">C</Radio.Item>
      </Radio>
    </Form.Item>
    <Form.Item name="isActive">
      <Button type="secondary">Submit</Button>
    </Form.Item>
  </Form>
)
