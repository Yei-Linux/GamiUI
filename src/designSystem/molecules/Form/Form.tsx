import React, { useEffect } from 'react'
import FormProvider from '../../../context/FormProvider'
import context from '../../../context/FormProvider/context'
import { IGeneralProps } from '../../../core/domain/interfaces/IGeneralProps'
import useCloneElement from '../../../hooks/useCloneElements'
import useStore from '../../../hooks/useStore'
import Button from '../../atoms/Button'
import Input from '../../atoms/Input'
import Number from '../../atoms/Number'
import Password from '../../atoms/Password'
import Radio from '../../atoms/Radio'
import Select from '../../atoms/Select'
import Switch from '../../atoms/Switch'
import TextArea from '../../atoms/TextArea'

import { FormItemWrapper, FormLabel, FormWrapper } from './Form.styles'

export interface FormProps extends IGeneralProps {
  onSubmitForm?: any
  children: React.ReactNode
}

const FormContent = ({ children, onSubmitForm = null, ...args }: FormProps) => {
  const { setCallbacks } = useStore({context})

  useEffect(() => {
    setCallbacks({
      onFinish: (values: any) => {
        onSubmitForm(values)
      },
    })
  }, [])

  return <FormWrapper {...args}>{children}</FormWrapper>
}

const Form = ({ children, onSubmitForm, ...args }: FormProps) => {
  return (
    <FormProvider>
      <FormContent onSubmitForm={onSubmitForm} {...args}>
        {children}
      </FormContent>
    </FormProvider>
  )
}

Form.Item = ({
  label = '',
  name,
  children,
}: {
  label?: string
  name: string
  children: React.ReactNode
}) => {
  const { formValue, setFormValues, onClickSubmit } = useStore({context})

  const handleChangeValue = (value: any): void => {
    setFormValues({ name, value })
  }

  const submitValues = () => onClickSubmit(formValue)

  const { validatorChildrenLength, childrenWithProps } = useCloneElement({
    children,
    propsElement: {
      name,
      value: formValue?.[name] ?? '',
      onChangeFormItem: handleChangeValue,
    },
    maxChildrenNumber: 1,
    advancedOptions: {
      propsOfElement: [
        { props: { onClick: submitValues }, childrenConditionTypes: [Button] },
      ],
    },
    childrenTypes: [
      Input,
      Password,
      TextArea,
      Select,
      Radio,
      Number,
      Switch,
      Button,
    ],
  })

  if (validatorChildrenLength(childrenWithProps)) {
    return null
  }

  return (
    <FormItemWrapper>
      {label != '' && (
        <FormLabel
          fontWeight="SEMIBOLD"
          text={label}
          width="NONE"  
          heigth="NORMAL"
          border="NONE"
        />
      )}
      {childrenWithProps}
    </FormItemWrapper>
  )
}

Form.defaultProps = {}

export default Form
