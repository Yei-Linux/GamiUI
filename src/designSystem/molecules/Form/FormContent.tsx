import React, { useEffect } from 'react'

import useFormStore from '../../../hooks/store/useFormStore'
import useGenerateYupSchema from '../../../hooks/useGenerateYupSchema'
import useYupValidation, { TYupSchema } from '../../../hooks/useYupValidation'
import { IFormProps } from './Form'

import * as S from './Form.styles'

const FormContent = ({
  children,
  onSubmitForm = null,
  ...args
}: IFormProps) => {
  const { yupSchema, formValue, setCallbacks, setYupSchema, onClickSubmit } =
    useFormStore()
  const { buildingYupSchema } = useGenerateYupSchema({ children })

  const { validate } = useYupValidation({ schema: yupSchema as TYupSchema })

  useEffect(() => {
    setCallbacks({
      onFinish: (values: any) => {
        onSubmitForm(values)
      },
    })

    const yupSchema = buildingYupSchema()
    setYupSchema(yupSchema)
  }, [])

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault()

    const error = validate(formValue)

    if (error) return

    onClickSubmit(formValue)
  }

  return (
    <S.Form onSubmit={handleSubmit} {...args}>
      {children}
    </S.Form>
  )
}

export default FormContent
