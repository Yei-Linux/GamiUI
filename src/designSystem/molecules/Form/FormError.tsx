import React, { Fragment } from 'react'
import useFormStore from 'hooks/store/useFormStore'
import Icon from 'designSystem/atoms/Icon'

import * as S from './Form.styles'

export interface IFormError {
  name: string
}

const FormError = ({ name }: IFormError) => {
  const { yupErrors } = useFormStore()

  return (
    <Fragment>
      {yupErrors?.[name] && (
        <S.FormError>
          <Icon name="errorForm" size="15px" fill="#E32B2B" />
          <S.FormErrorLabel text={yupErrors?.[name]} />
        </S.FormError>
      )}
    </Fragment>
  )
}

export default FormError
