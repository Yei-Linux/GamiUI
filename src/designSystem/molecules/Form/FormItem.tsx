import React from 'react';
import useFormStore from '../../../hooks/store/useFormStore';
import useCloneElement from '../../../hooks/useCloneElements';

import * as S from './Form.styles'

import Button from '../../atoms/Button'
import Input from '../../atoms/Input'
import Number from '../../atoms/Number'
import Password from '../../atoms/Password'
import Radio from '../../atoms/Radio'
import Select from '../../atoms/Select'
import Switch from '../../atoms/Switch'
import TextArea from '../../atoms/TextArea'
import FormError from './FormError';

export type TRulesType = 'required' | 'email' | 'maxNumber' | 'minNumber'

export interface IRules {
  type: TRulesType
  message: string
  value?: number
}

export interface IFormItem {
  label?: string
  name: string
  children: React.ReactNode
  rules?: IRules[]
}
 
const FormItem = ({ label = '', name, children }: IFormItem) => {
    const { formValue, setFormValues } = useFormStore()
  
    const handleChangeValue = (value: any): void => {
      setFormValues({ name, value })
    }
  
    const { validatorChildrenLength, childrenWithProps } = useCloneElement({
      children,
      propsElement: {
        name,
        value: formValue?.[name] ?? '',
        onChangeFormItem: handleChangeValue,
      },
      maxChildrenNumber: 1,
      advancedOptions: {
        propsOfElement: [{ props: {}, childrenConditionTypes: [Button] }],
      },
      childrenTypes: [
        Input,
        Password,
        TextArea,
        Select,
        Radio,
        Number,
        Switch,
        Button,
      ],
    })
  
    if (validatorChildrenLength(childrenWithProps)) {
      return null
    }
  
    return (
      <S.FormItem>
        {label != '' && (
          <S.FormLabel
            fontWeight="SEMIBOLD"
            text={label}
            width="NONE"
            heigth="NORMAL"
            border="NONE"
          />
        )}
        {childrenWithProps}
        <FormError name={name} />
      </S.FormItem>
    )
  }
 
export default FormItem;

