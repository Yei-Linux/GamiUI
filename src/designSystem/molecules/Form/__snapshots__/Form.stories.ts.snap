// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Molecules/Form Basic 1`] = `
<div
  className="css-pnvx7p-FormWrapper enokt0p0"
>
  <div
    className="css-o5sgrd-FormItemWrapper enokt0p1"
  >
    <div
      className="enokt0p2 css-yckgou-RichTextWrapper-FormLabel e17e5g740"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Nombres</p>
",
        }
      }
      width="NONE"
    />
    <input
      autoComplete="off"
      className="css-gwomp1-InputContainer e7vlt9n1"
      name="names"
      onChange={[Function]}
      placeholder="Ingresa tus nombres"
      prefix={null}
      readOnly={false}
      type="input"
      value=""
      width="NORMAL"
    />
  </div>
  <div
    className="css-o5sgrd-FormItemWrapper enokt0p1"
  >
    <div
      className="enokt0p2 css-yckgou-RichTextWrapper-FormLabel e17e5g740"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Contraseña</p>
",
        }
      }
      width="NONE"
    />
    <div
      autoComplete="off"
      className="css-11c3v5m-InputWrapper e7vlt9n0"
      name="password"
      placeholder="Ingresa tu contraseña"
      prefix={
        <div
          onClick={[Function]}
        >
          <Icon
            fill="blue"
            name="closeeye"
            size="30px"
          />
        </div>
      }
      readOnly={false}
      type="password"
      value=""
      width="NORMAL"
    >
      <div
        onClick={[Function]}
      >
        <i
          className="css-6tw2qr-IconWrapper e1cr9qb30"
        >
          <svg
            fill="blue"
            height="30px"
            viewBox="0 0 512.52 512.52"
            width="30px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <g>
                <path
                  d="m512.52 187.167-3.34 6.69c-.71 1.45-18.07 35.74-57.99 69.95-23.36 20.02-49.72 35.97-78.35 47.4-35.65 14.24-74.87 21.46-116.58 21.46s-80.93-7.22-116.58-21.46c-28.63-11.43-54.99-27.38-78.35-47.4-39.92-34.21-57.28-68.5-57.99-69.95l-3.34-6.69 3.34-6.69c.71-1.44 18.07-35.74 57.99-69.95 23.36-20.01 49.72-35.96 78.35-47.4 35.65-14.24 74.87-21.46 116.58-21.46s80.93 7.22 116.58 21.46c28.63 11.44 54.99 27.39 78.35 47.4 39.92 34.21 57.28 68.51 57.99 69.95z"
                  fill="#fbf3f3"
                />
                <path
                  d="m512.52 187.167-3.34 6.69c-.71 1.45-18.07 35.74-57.99 69.95-23.36 20.02-49.72 35.97-78.35 47.4-35.65 14.24-74.87 21.46-116.58 21.46v-291c41.71 0 80.93 7.22 116.58 21.46 28.63 11.44 54.99 27.39 78.35 47.4 39.92 34.21 57.28 68.51 57.99 69.95z"
                  fill="#e8e3f0"
                />
                <ellipse
                  cx="256.26"
                  cy="187.167"
                  fill="#adcfff"
                  rx="114.12"
                  ry="114.12"
                  transform="matrix(.707 -.707 .707 .707 -57.291 236.023)"
                />
                <path
                  d="m370.38 187.167c0 63.03-51.09 114.12-114.12 114.12v-228.24c63.03 0 114.12 51.101 114.12 114.12z"
                  fill="#6aa8ff"
                />
                <ellipse
                  cx="256.26"
                  cy="189.437"
                  fill="#363671"
                  rx="59.42"
                  ry="59.42"
                  transform="matrix(.257 -.967 .967 .257 7.395 388.49)"
                />
                <path
                  d="m315.68 189.437c0 32.82-26.6 59.42-59.42 59.42v-118.84c32.82 0 59.42 26.6 59.42 59.42z"
                  fill="#16164d"
                />
                <circle
                  cx="298.311"
                  cy="151.391"
                  fill="#fbf3f3"
                  r="28.877"
                />
              </g>
              <ellipse
                cx="412.11"
                cy="370.703"
                fill="#ff7c48"
                rx="100.15"
                ry="100.15"
                transform="matrix(.707 -.707 .707 .707 -141.422 399.982)"
              />
              <path
                d="m512.26 370.703c0 55.22-44.93 100.15-100.15 100.15v-200.3c55.22 0 100.15 44.92 100.15 100.15z"
                fill="#ff3939"
              />
              <path
                d="m412.11 355.707 10 14.996-10 15.004h-41.48v-30z"
                fill="#fbf3f3"
              />
              <path
                d="m412.11 355.707h41.48v30h-41.48z"
                fill="#e8e3f0"
              />
            </g>
          </svg>
        </i>
      </div>
      <input
        autoComplete="off"
        className="css-7dwcir-InputContainer e7vlt9n1"
        name="password"
        onChange={[Function]}
        placeholder="Ingresa tu contraseña"
        prefix={
          <div
            onClick={[Function]}
          >
            <Icon
              fill="blue"
              name="closeeye"
              size="30px"
            />
          </div>
        }
        readOnly={false}
        type="password"
        value=""
        width="NORMAL"
      />
    </div>
  </div>
  <div
    className="css-o5sgrd-FormItemWrapper enokt0p1"
  >
    <div
      className="enokt0p2 css-yckgou-RichTextWrapper-FormLabel e17e5g740"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Celular</p>
",
        }
      }
      width="NONE"
    />
    <input
      autoComplete="off"
      className="css-gwomp1-InputContainer e7vlt9n1"
      name="phone"
      onChange={[Function]}
      placeholder="Ingresa tu celular"
      prefix={null}
      readOnly={false}
      type="input"
      value=""
      width="NORMAL"
    />
  </div>
  <div
    className="css-o5sgrd-FormItemWrapper enokt0p1"
  >
    <div
      className="enokt0p2 css-yckgou-RichTextWrapper-FormLabel e17e5g740"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Comentario</p>
",
        }
      }
      width="NONE"
    />
    <textarea
      className="css-155tpye-TextAreaWrapper eieaaae0"
      cols={5}
      name="comment"
      onChange={[Function]}
      placeholder="Ingresa tu comentario"
      rows={5}
      value=""
      width="NORMAL"
    />
  </div>
  <div
    className="css-o5sgrd-FormItemWrapper enokt0p1"
  >
    <div
      className="enokt0p2 css-yckgou-RichTextWrapper-FormLabel e17e5g740"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Activar Notificaciones</p>
",
        }
      }
      width="NONE"
    />
    <button
      className="css-13kjyr7-SwitchWrapper egh0tog0"
      name="isActive"
      onClick={[Function]}
      value=""
      width="NORMAL"
    >
      <div
        className="css-1lmkk7b-SwitchBall egh0tog1"
      />
    </button>
  </div>
  <div
    className="css-o5sgrd-FormItemWrapper enokt0p1"
  >
    <div
      className="enokt0p2 css-yckgou-RichTextWrapper-FormLabel e17e5g740"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Elige un pais</p>
",
        }
      }
      width="NONE"
    />
    <div
      className="css-htg7a1-SelectWrapper ee0eyh30"
    >
      <div
        className="css-1dhedq7-Selectinput ee0eyh31"
        onClick={[Function]}
      >
        <input
          autoComplete="off"
          className="css-gwomp1-InputContainer e7vlt9n1"
          name="country"
          onChange={[Function]}
          placeholder="Type your option"
          prefix={null}
          readOnly={true}
          type="input"
          value=""
          width="NORMAL"
        />
      </div>
      <div
        className="css-ae753p-SelectOptionsWrapper ee0eyh32"
      >
        <div
          className="css-u0k5sv-OptionWrapper ee0eyh33"
          onClick={[Function]}
        >
          Option1
        </div>
        <div
          className="css-u0k5sv-OptionWrapper ee0eyh33"
          onClick={[Function]}
        >
          Option2
        </div>
        <div
          className="css-u0k5sv-OptionWrapper ee0eyh33"
          onClick={[Function]}
        >
          Option3
        </div>
      </div>
    </div>
  </div>
  <div
    className="css-o5sgrd-FormItemWrapper enokt0p1"
  >
    <div
      className="enokt0p2 css-yckgou-RichTextWrapper-FormLabel e17e5g740"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Elige tu genero</p>
",
        }
      }
      width="NONE"
    />
    <div
      className="css-k89t9i-RadioGroupWrapper eldjpxi0"
      name="gender"
      value=""
      width="NORMAL"
    >
      <label
        className="css-1ur0gxt-RadioContainer eldjpxi1"
      >
        <span
          className="css-1c1trwd-RadioCheck eldjpxi2"
        >
          <input
            className="css-uckndt-RadioInput eldjpxi3"
            onClick={[Function]}
          />
          <span
            className="css-egwf0x-RadioInner eldjpxi4"
          />
        </span>
        <span
          className="css-26doda-RadioText eldjpxi5"
        >
          A
        </span>
      </label>
      <label
        className="css-1ur0gxt-RadioContainer eldjpxi1"
      >
        <span
          className="css-1c1trwd-RadioCheck eldjpxi2"
        >
          <input
            className="css-uckndt-RadioInput eldjpxi3"
            onClick={[Function]}
          />
          <span
            className="css-egwf0x-RadioInner eldjpxi4"
          />
        </span>
        <span
          className="css-26doda-RadioText eldjpxi5"
        >
          B
        </span>
      </label>
      <label
        className="css-1ur0gxt-RadioContainer eldjpxi1"
      >
        <span
          className="css-1c1trwd-RadioCheck eldjpxi2"
        >
          <input
            className="css-uckndt-RadioInput eldjpxi3"
            onClick={[Function]}
          />
          <span
            className="css-egwf0x-RadioInner eldjpxi4"
          />
        </span>
        <span
          className="css-26doda-RadioText eldjpxi5"
        >
          C
        </span>
      </label>
    </div>
  </div>
  <div
    className="css-o5sgrd-FormItemWrapper enokt0p1"
  >
    <button
      className="css-1ysm1ze-ButtonWrapper ekfoxpm0"
      onClick={[Function]}
      type="submit"
      width="SMALL"
    >
      Submit
    </button>
  </div>
</div>
`;
